ui/src/app/machines/views/AddMachine/AddMachineForm/AddMachineForm.js

```diff
@@ -0,0 +1,224 @@
+  if (powerType && powerType.fields) {
+    powerType.fields.forEach(field => {
+      trimmed[field.name] = powerParameters[field.name];
+    });
+  }
+  return trimmed;
+};
+
+const generateMachineSchema = parametersSchema =>
+  Yup.object().shape({
+    architecture: Yup.string().required("Architecture required"),
+    domain: Yup.string().required("Domain required"),
+    extra_macs: Yup.array().of(
+      Yup.string().matches(
+        /^([0-9A-Fa-f]{2}:){5}([0-9A-Fa-f]{2})$/,
+        "Invalid MAC address"
+      )
+    ),
+    hostname: Yup.string(),
+    min_hwe_kernel: Yup.string(),
+    pool: Yup.string().required("Resource pool required"),
+    power_parameters: Yup.object().shape(parametersSchema),
+    power_type: Yup.string().required("Power type required"),
+    pxe_mac: Yup.string()
+      .matches(/^([0-9A-Fa-f]{2}:){5}([0-9A-Fa-f]{2})$/, "Invalid MAC address")
+      .required("At least one MAC address required"),
+    zone: Yup.string().required("Zone required")
+  });
+
+export const AddMachineForm = () => {
+  const dispatch = useDispatch();
+
+  const architectures = useSelector(generalSelectors.architectures.get);
+  const architecturesLoaded = useSelector(
+    generalSelectors.architectures.loaded
+  );
+  const defaultMinHweKernel = useSelector(
+    generalSelectors.defaultMinHweKernel.get
+  );
+  const defaultMinHweKernelLoaded = useSelector(
+    generalSelectors.defaultMinHweKernel.loaded
+  );
+  const domains = useSelector(domainSelectors.all);
+  const domainsLoaded = useSelector(domainSelectors.loaded);
+  const hweKernelsLoaded = useSelector(generalSelectors.hweKernels.loaded);
+  const machineSaved = useSelector(machineSelectors.saved);
+  const machineSaving = useSelector(machineSelectors.saving);
+  const machineErrors = useSelector(machineSelectors.errors);
+  const powerTypes = useSelector(generalSelectors.powerTypes.get);
+  const powerTypesLoaded = useSelector(generalSelectors.powerTypes.loaded);
+  const resourcePools = useSelector(resourcePoolSelectors.all);
+  const resourcePoolsLoaded = useSelector(resourcePoolSelectors.loaded);
+  const zones = useSelector(zoneSelectors.all);
+  const zonesLoaded = useSelector(zoneSelectors.loaded);
+
+  const [savingMachine, setSavingMachine] = useState(false);
+  const [powerType, setPowerType] = useState("");
+  const [MachineSchema, setMachineSchema] = useState(generateMachineSchema({}));
+
+  // Fetch all data required for the form.
+  useEffect(() => {
+    dispatch(domainActions.fetch());
+    dispatch(generalActions.fetchArchitectures());
+    dispatch(generalActions.fetchDefaultMinHweKernel());
+    dispatch(generalActions.fetchHweKernels());
+    dispatch(generalActions.fetchPowerTypes());
+    dispatch(resourcePoolActions.fetch());
+    dispatch(zoneActions.fetch());
+  }, [dispatch]);
+
+  // When selected power type changes, update the validation schema with new
+  // power_parameters shape.
+  useEffect(() => {
+    if (powerType) {
```

Comment:
Do we also need to unset the schema if this gets set back to null or can't that happen?

Reply:
It shouldn't be possible to unset the power type once one is selected, but it doesn't hurt to add that check in anyway.

Fixed.